File: Java/soln-recursion-problems/generate-parentheses.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/generate-parentheses/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    List<String> result=new ArrayList<>();
    
    public List<String> generateParenthesis(int n) {
        if(n==0)
           return result;
        solve(n,n,"");
        return result;        
    }
    
    public void solve(int open_br, int close_br, String str){
        if(open_br==0 && close_br==0){
            result.add(str);
            return; 
        }    
        if(open_br!=0)
            solve(open_br-1,close_br,str+"(");

        if(close_br>open_br){
            solve(open_br,close_br-1,str+")");
        }
    }
}

File: Python/soln-recursion-problems/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res=[]
        
        if n==0:
            return res
        
        def solve(open_br, close_br, str):
            
            if open_br==close_br and open_br==0:
                res.append(str)
                return
            
            if open_br>0:
                solve(open_br-1,close_br,str+"(")
            
            if close_br>open_br:
                solve(open_br,close_br-1,str+")")
                
        solve(n,n,"")
        return res